### DISCUSSION TITLE
Layout Manager

### DISCUSSION THREAD

**Discussion by ShaMan123** [fabricjs/fabric.js#9148] [General]
## Motivation 
As we discussed we want to extract the layout logic from Group to a standalone manager.
This is done for a number of reasons:

- scoping
- ease of customization

## Spec

- declarative simple lifecycle
- easy to override and to hook into
- imperative triggering of layout
- fine garined control over each layout cycle - passing enough context between steps
- different layouts
- future proof
   - implenemtaion should prepare the ground for related features (css layout, html rendering etc.)
   - designed in such a way that heavy customization can be accomplished without needing to patch the codebase.

## Progress

#9152 extracts the logic to a manager and reshaping it according to this discussion
I will also update this description accordingly.

**Comment by ShaMan123** [comment]
@asturur @melchiar @iocoker @ebigtm @jiayihu I would like your input and insights so we come up with a robust solution.

**Reply by ShaMan123** [reply]
@ebigtm

**Comment by ShaMan123** [comment]
What about persistance? Should we serialize something or leave it to the dev?

**Reply by asturur** [reply]
Persistance will be complicated indeed, i was thinking about that a week or two ago.
a group with different layout gets rendered differently of another group, so the layout is part of the serialization.
At the same time a layout manger is added in the constructor and so i don't really know what we can do.

**Comment by ebigtm** [comment]
@ShaMan123 as mentioned before I really like the idea, IMO it should take into account two main aspects which we find ourselves struggeling with in almost every move., both are related to HTML dom elements, maybe it needs to be sperate utils?
1. The css aspect, this approach should make it easy to maintanin with furture css layout implemetations and edge usecases.
2. Might be not super related in first glimps - The "HTML rendering" side would be amazing to have, for example text rendering in fabric is very lacking and to be able to take the structure of a whole paragraph of text with UL, LI, and more would save us so much development time.

However, I would start, as you said, with the basic structure not implement "final products", just the methods that can be easiliy expanded

**Reply by asturur** [reply]
regariding point number 2, which are other canvas based libraries that have a better text implementation we can take inspiration from?

**Reply by ShaMan123** [reply]
I was thinking...
How could we make the layout manager suitabke for text layout as well.
That would be awesome.
Is there a way in html to measure the text? That will make our lives easier.

You said you saw pixi.js use our text and they are a good repo....

**Reply by ShaMan123** [reply]
We can try to use logic from a html to image repo...

**Reply by asturur** [reply]
html measures text with the api of canvas or with the layout of bounding box of a div containing the text.

**Reply by asturur** [reply]
this layout manager is for group only. Its job should be to take out from the group class measuring and dimension of the group and objects and allow other people to write their own group layouting. It should be a trasposition of the current code in a more focused way, possibly removing code and not doing other new things.
i m begging you

**Reply by ebigtm** [reply]
I agree even after almost 3 years of fiddiling around I just realize now stuff that I should have realized a while back so documentation is crucial.

A good documentation with examples should do the job.

Regarding text not quite sure what you meant but we ended up inventing the wheel again to be able to present complicated text (WYSIWYG text with different styles, and different bullets) but maybe it is still connected to groups as we can think of what i just said as a group with a basic structure by sizes and paragraphs...

**Reply by ShaMan123** [reply]
> this layout manager is for group only. Its job should be to take out from the group class measuring and dimension of the group and objects and allow other people to write their own group layouting. It should be a trasposition of the current code in a more focused way, possibly removing code and not doing other new things.
> i m begging you

Think of a case in which I want to layout some text and some inline shape.
Now what?
Should we reinvent another mechanism?
HTML has one, css, right?

It does not have to be now, but I think we should plan for it. Think ahead instead of reaching a point where we are completely limited by the current solution. Writing something and then tossing it.

**Reply by asturur** [reply]
but this is not planning for it.

**Comment by jiayihu** [comment]
In this case I think it could be good to adopt some OOP pattern, like the Strategy one: https://refactoring.guru/design-patterns/strategy to allow easier customisation or plugging custom layout managers into fabric, like the DOM ones. 
In this way cool Layout Managers like CSS Grid and Flexbox equivalent can be opt-in and without increasing the bundle size.

Another source of inspiration for the API would be the CSS Houdini Layout API, which allows to define your own layout behaviour into CSS, e.g. Masonry: https://github.com/w3c/css-houdini-drafts/blob/main/css-layout-api/EXPLAINER.md. For instance, in the documentation they mention Text layouting as well, so that would in theory allow to use maybe the same Layout Manager for both Groups and text rendering as asked by @ebigtm

**Comment by Lazauya** [comment]
I was about to make a new post when I remembered this discussion. I don't want to dirty up this thread so tell me if this is too out of scope or irrelevant.

I was working on a project recently and got really lost by the current group behavior. My code was similar to this:
```
const greenRect = new Rect({
  width: 100,
  height: 100,
  fill: "#88FF00",
  strokeWidth: 0
});

const blueRect = new Rect({
  width: 100,
  height: 100,
  left: 100,
  top: 100,
  fill: "#0088FF",
  strokeWidth: 0
});

const pinkRect = new Rect({
  width: 250,
  height: 250,
  fill: "#FF0088",
  strokeWidth: 0
});

const child = new Group([pinkRect]);
const parent = new Group([child, greenRect, blueRect]);
canvas.add(parent);
```
It draws the things as expected, no problem there. But when you look at the `left` and `top` attributes, things become way more confusing. `greenRect.top` is `-125`! This feels incredibly unintuitive.

It makes sense to transform all the children into the same coordinate space regardless of the group's `originX` or `originY` (if I understand correctly it uses `parent`'s center for it's "internal" origin). My issue is that I think that changing the meaning of the consumer-facing variables like `left` and `top` when an object is in a group makes it sooo much more confusing from a beginner perspective. I feel like we need API's or properties for the group specific position/offset, and `left` and `top` should reflect the actual canvas-space coordinates of the object, regardless of which group it's in. I'm curious what other's thoughts are.

**Reply by ShaMan123** [reply]
I agree with most of your comment.
left, top are legacy props. I want them renamed to x, y.
I have exposed getXY, which is what you are after.
However, object props must be relative to the parent. Methods can return whatever.

**Reply by asturur** [reply]
i just want to clarify that left and top are not legacy.
Switching top and left to x and y is an incredible pain for every existing project.

It is true that  having the group with origin based on the group center is a bit confusing and that we would like everything always ready in the space we want. but keeping in sync 2 sets of coordinates is painful as well.
Probably the error was to have the canvas based from the top left corner, should have been from the center as well.

**Reply by ShaMan123** [reply]
It is too confusing.
I rememeber it as the first thing I bashed my ahead against. It must be documeneted.
In the redo coords pr I think we can manage to migrate from it while keeping it as legacy and not breaking.

**Reply by asturur** [reply]
that giant pr for coords shouldn't be the base for plans, i would like a detailed plan for coords ans what we think it should be done, thing i didn't see.

if every change is approached with 'i should change also this and that' is not helpful.

Again 'too complicated for me' is not a metric. let's write documentation not more code

**Reply by ShaMan123** [reply]
Please, let's plan. I am wishing that as well.

**Comment by ShaMan123** [comment]
The pr has matured, reviews and thoughts are welcome.
Reading this discussiion's description, seems it landed the spec.
